passes/sat/qbfsat.cc

```diff
@@ -0,0 +1,522 @@
+			unsat_regex_found = true;
+	}
+#ifndef NDEBUG
+	log_assert(!sol.unknown && sol.sat? sat_regex_found : true);
+	log_assert(!sol.unknown && !sol.sat? unsat_regex_found : true);
+#endif
+}
+
+std::map<std::string, std::string> get_hole_loc_name_map(RTLIL::Module *module, const QbfSolutionType &sol) {
+	std::map<std::string, std::string> hole_loc_to_name;
+	for (auto cell : module->cells()) {
+		std::string cell_src = cell->get_src_attribute();
+		auto pos = sol.hole_to_value.find(cell_src);
+		if (pos != sol.hole_to_value.end()) {
+#ifndef NDEBUG
+			log_assert(cell->type.in("$anyconst", "$anyseq"));
+			log_assert(cell->hasPort(ID::Y));
+			log_assert(cell->getPort(ID::Y).is_wire());
+#endif
+			hole_loc_to_name[pos->first] = cell->getPort(ID::Y).as_wire()->name.str();
+		}
+	}
+
+	return hole_loc_to_name;
+}
+
+void write_solution(RTLIL::Module *module, const QbfSolutionType &sol, const std::string &file) {
+	std::ofstream fout(file.c_str());
+	if (!fout)
+		log_cmd_error("could not open solution file for writing.\n");
+
+	std::map<std::string, std::string> hole_loc_to_name = get_hole_loc_name_map(module, sol);
+	for(auto &x : sol.hole_to_value)
+		fout << hole_loc_to_name[x.first] << "=" << x.second << std::endl;
+}
+
+void specialize_from_file(RTLIL::Module *module, const std::string &file) {
+	YS_REGEX_TYPE hole_assn_regex = YS_REGEX_COMPILE_WITH_SUBS("^(.*)=([01]+)$");
+	YS_REGEX_MATCH_TYPE m;
+	std::ifstream fin(file.c_str());
+	if (!fin)
+		log_cmd_error("could not read solution file.\n");
+
+	std::string buf;
+	while (std::getline(fin, buf)) {
+		log_assert(YS_REGEX_NS::regex_search(buf, m, hole_assn_regex));
+		std::string hole_name = m[1].str();
+		std::string hole_value = m[2].str();
+
+		RTLIL::Wire *wire = module->wire(hole_name);
+#ifndef NDEBUG
+		log_assert(wire != nullptr);
+		log_assert(wire->width > 0 && hole_value.size() == static_cast<unsigned long>(wire->width));
+#endif
+
+		log("Specializing %s from file with %s = %d'b%s.\n", module->name.c_str(), hole_name.c_str(), wire->width, hole_value.c_str());
+		std::vector<RTLIL::SigBit> value_bv;
+		value_bv.reserve(wire->width);
+		for (char c : hole_value)
+			value_bv.push_back(c == '1'? RTLIL::S1 : RTLIL::S0);
+		std::reverse(value_bv.begin(), value_bv.end());
+		module->connect(wire, value_bv);
+	}
+}
+
+void specialize(RTLIL::Module *module, const QbfSolutionType &sol) {
+	std::map<std::string, std::string> hole_loc_to_name = get_hole_loc_name_map(module, sol);
+	for (auto &it : sol.hole_to_value) {
+		std::string hole_loc = it.first;
+		std::string hole_value = it.second;
+
+		auto pos = hole_loc_to_name.find(hole_loc);
+#ifndef NDEBUG
+		log_assert(pos != hole_loc_to_name.end());
+#endif
+
+		std::string hole_name = hole_loc_to_name[hole_loc];
+		RTLIL::Wire *wire = module->wire(hole_name);
+#ifndef NDEBUG
+		log_assert(wire != nullptr);
+		log_assert(wire->width > 0 && hole_value.size() == static_cast<unsigned long>(wire->width));
```

Comment:
By convention Yosys tends to use GetSize(hole_value) to get the size as an integer rather than casting either side manually